posts/2024-06-20-liberty-user-feature-tutorial.adoc

```diff
@@ -0,0 +1,252 @@
+
+Inside `<instructions>` add the line `<Bundle-Name>example.user.feature.human.name</Bundle-Name>` and `<Bundle-SymbolicName>example.user.feature.integration.machine.name</Bundle-SymbolicName>`.
+
+Also inside `<instructions>` you will find the tag `<Export-Package>`, populate it with.
+
+```
+${new.integration.code.api.package};version="1.0.0",
+${new.integration.code.private.package};version="1.0.0"
+```
+
+These classes will not be registered correctly without a version number.
+
+Going back to the parent `pom.xml` set these properties:
+
+```
+<new.integration.code.private.package>com.ibm.example.cdi.internal</new.integration.code.private.package>    <new.integration.code.api.package>com.ibm.example.cdi.api</new.integration.code.api.package>
+```
+
+==== Step two: Create the ESA.
+
+Open Liberty features are packaged as Enterprise Subsystem Archive.  We will create one that will include both our new integration code and the pre-existing library.
+
+Open `esa/pom.xml`
+
+The first thing we need to do is ensure our ESA will have a `manifest.mf `file. Set `<generateManifest>true</generateManifest>` in the configuration of `esa-maven-plugin`.
+
+Now in instructions we will set a subystem symbolic name `<Subsystem-SymbolicName>example.user.feature.esa.machine.name;visibility:=public</Subsystem-SymbolicName>`. Seting the visibility is required.
+
+We will also need an IBM shortname. Add `<IBM-ShortName>${feature.name}</IBM-ShortName>` inside instructions, and set ${feature.name} to `example-feature` in the root pom.xml
+
+Finally add
+
+```
+${pre.existing.library.package};version="3.14.0",
+${new.integration.code.api.package};version="1.0.0"
+```
+
+Under IBM-API-Package. This will make those two packages visible to applications at runtime.
+
+The ESA is now complete. But there is one final step, set `${pre.existing.library.package}` to `org.apache.commons.lang3.concurrent` in the parent pom.xml
+
+==== Step three: Create the Bill of Materials.
+
+The `liberty-maven-plugin` requires a bill of materials to find and install features. In the real world the Bill of Materials may be defined in the ESA’s pom.xml, but for clarity this tutorial will keep them separate.
+
+Open bom/pom.xml and add the following dependency.
+
+```
+      <dependency>
+        <groupId>com.ibm.example.user.feature</groupId>
+        <!-- This is ${esa.artefact.id}. A variable cannot be used here -->
+        <!-- As this needs to be readable outside this project. -->
+        <artifactId>liberty-feature</artifactId>
+        <version>1.0-SNAPSHOT</version>
+        <type>esa</type>
+        <scope>provided</scope>
+      </dependency>
+```
+
+==== step four, add your liberty-feature to a liberty server.
+
+Go to demo/src and add the line `<feature>usr:example-feature</feature>` inside `<featureManager>`. The format is `usr:${IBM-ShortName}` but naturally a liberty server.xml cannot use properties from a pom.xml
```

Comment:
an example might be helpful here too, as it's not immediately clear what the relationship is between `usr:example-feature` and `usr:${IBM-ShortName}` is. Can you use a consistent example variable here if `example-feature` and  `IBM-ShortName` are meant to be the same string?